openfisca_core/simulation_builder.py

```diff
@@ -159,9 +165,37 @@ def build_default_simulation(self, tax_benefit_system, count = 1, **kwargs):
             entity.ids = np.array(range(count))
             if not entity.is_person:
                 entity.members_entity_id = entity.ids  # Each person is its own group entity
-                entity.members_role = entity.filled_array(entity.flattened_roles[0])
         return simulation
 
+    def create_entities(self, tax_benefit_system):
+        self.entities_instances = tax_benefit_system.instantiate_entities()
+
+    def declare_person_entity(self, person_singular, persons_ids: Iterable):
+        person_instance = self.entities_instances[person_singular]
+        person_instance.ids = np.array(list(persons_ids))
+        person_instance.count = len(person_instance.ids)
+
+        self.persons_plural = person_instance.plural
+
+    def declare_entity(self, entity_singular, entity_ids: Iterable):
+        entity_instance = self.entities_instances[entity_singular]
+        entity_instance.ids = np.array(list(entity_ids))
+        entity_instance.count = len(entity_instance.ids)
+        return entity_instance
+
+    def nb_persons(self, entity_singular, role = None):
+        return self.entities_instances[entity_singular].nb_persons(role = role)
+
+    def join_with_persons(self, group_instance, group_to_person_projection, roles: Iterable[str] = None):
+        group_sorted_indices = np.unique(group_to_person_projection, return_inverse = True)[1]
+        group_instance.members_entity_id = np.argsort(group_instance.ids)[group_sorted_indices]
+
+        if roles is not None:
+            group_instance.members_role = [group_instance.get_role(role_name) for role_name in roles]
```

Comment:
@fpagnoux I don't think we have a dire need to optimize this as it's a one-time setup cost. The `np.select` version strikes me as less readable (maybe because I'm used to seeing it in the context of formulas, doing a different job). I'll do a quick performance measurement to assess the trade-off.

Reply:
Unless I've misunderstood something, the `np.select` approach is actually a lot slower.

```
>>> flattened_roles = ['b','c','a']
>>> roles = np.repeat('a',10000000)
>>> start = time.time(); indices = np.select([roles == role for role in flattened_roles], flattened_roles); duration = time.time() - start
>>> duration
33.49795985221863
>>> start = time.time(); indices = [flattened_roles.index(role_name) for role_name in roles]; duration = time.time() - start
>>> duration
9.044532775878906
```

Reply:
NumPy is vectorial computation library, meaning it allows to make calculations on large arrays faster than Python native loops.  The benefits are typically visible on _large_ arrays.  For small arrays, the constant costs of NumPy operations usually outweight the benefits of the vectorial approach.

Reply:
Using a script that is more similar to what happens in the code, this is what I get:

```py
import numpy as np
import time

class Role(object):
  def __init__(self, name):
    self.name = name

flattened_roles = [Role('a'), Role('b'), Role('c')]

role_by_name = {role.name: role for role in flattened_roles}

# Randomly generate a vector of role name
flattened_roles_name = ['a', 'b', 'c']
role_names_array = np.random.choice(flattened_roles_name, 1000000)

# Generate the role object vector using numpy
start = time.time(); roles_array = np.select([role_names_array == role.name for role in flattened_roles], flattened_roles); duration_n = time.time() - start
print(duration_n)

# Generate the role object vector using a Python loop
start = time.time(); roles_array = [role_by_name[role_name] for role_name in role_names_array]; duration_p = time.time() - start
print(duration_p)

print(f'Python is {round(duration_p/duration_n, 2)} slower than NumPy')
```

Result:

```
0.05843806266784668
0.340930700302124
Python is 5.83 slower than NumPy
```

Reply:
> I don't think we have a dire need to optimize this as it's a one-time setup cost

The thing is a "one-time setup" is not _that_ different from a formula. In a typical simulation where I compute an annual `income_tax`, I'll also run the formula code only once. Yet there was always a strong push from economists like @benjello to avoid looping over any vector. 

I agree that the performance cost here is not a deal breaker, but it seems strange to me that we start breaking the well-established don't-loop-over-vector rule, so far consistently applied in the code.

Reply:
Hmm. The results I'm getting on my home Mac Mini are going the opposite way from what I reported from last night on my Macbook, whether it's on your script or my original script, so I'm not quite sure what is going on here. Possibly effects of GC on a smaller RAM allocation?

I've pushed the `np.select` version, but note that this also incurs the cost of transforming the Iterable that is passed in to join_with_persons into a Numpy array in the first place, which was not modeled in the above comparisons.

> well-established don't-loop-over-vector rule

It bothers me that we frame performance-impacting choices as "following rules", as opposed to "understanding what's going on *and then* empirically verifying the effect with benchmarks". 

We still don't have a clear understanding of the performance cost we're paying for the "garden of forking paths" effect where we compute every choice of every `np.select` or `np.where`, for instance.

Reply:
> It bothers me that we frame performance-impacting choices as "following rules", as opposed to "understanding what's going on and then empirically verifying the effect with benchmarks".

Sure, but if the alternative is to use developers time to make benchmarks _each time_ we are considering iterating on a vector, like we've done here, I'm not sure it's worth the pain. 

As for the avoiding "garden of forking paths", this is for sure a very promising idea, but would still need to accommodate the tabular data usecase.